mltensor-explainer.md

```diff
@@ -0,0 +1,423 @@
+  );
+
+  // Rent the other MLTensor out to WebGPU.
+  const tensorizedGpuBufferAfterInference = gpuDevice.importExternalBuffer(mlTensor2);
+
+  // Create a bind group for `tensorizedGpuBufferAfterInference`,
+  // create a command encoder, etc to feed `tensorizedGpuBufferAfterInference`
+  // into a GPU shader which may blur the frame or replace background sections
+  // and then render the result
+  // ...
+
+  gpuDevice.queue.submit([texturizeAndRenderCommandEncoder.finish()]);
+
+  // Return the buffer to WebNN for the next frame.
+  tensorizedGpuBufferAfterInference.destroy();
+
+  // Call this method for each frame.
+  video.requestVideoFrameCallback(applyEffectToFrame);
+}
+```
+
+## Design Discussion
+
+### Timelines
+
+WebNN uses a programming model similar to [WebGPU's](https://www.w3.org/TR/webgpu/#programming-model), in that compute tasks are posted to a timeline - which I've referred to as an "ML context timeline" throughout this document - separate from the content timeline (i.e. "script"). See [the WebGPU documentation of timelines](https://www.w3.org/TR/webgpu/#programming-model-timelines) for more details.
+
+Specifying WebNN timelines is tracked in [#529](https://github.com/webmachinelearning/webnn/issues/529).
+
+### Device Affinity and Relationship to a `GPUDevice`
+
+The WebNN API requires the developer to declare how an `MLTensor` will be used (via `MLTensorUsageFlags`), which the user agent may use as a hint in deciding where to allocate the memory backing an `MLTensor`. Where the memory is ultimately allocated is up to the user agent.
+
+For example [an `MLContext` may be created with a `GPUDevice`](https://www.w3.org/TR/webnn/#dom-ml-createcontext-gpudevice), and creating an `MLTensor` from this context with the `MLTensorUsage.WEBGPU_INTEROP` flag expresses a clear intention to share the tensor with the given `GPUDevice`. However, there is no guarantee that sharing this tensor with WebGPU will be zero-copy.
+
+The `MLTensorUsage.READ_FROM` and `MLTensorUsage.WRITE_TO` flags likewise are hints to the user agent indicating that the underlying data will be read and written to, respectively, by script.
+
+### Importing an `MLTensor` to WebGPU
+
+Any `MLTensor` created with the `MLTensorUsage.WEBGPU_INTEROP` flag may be imported into any `GPUDevice`, though cross-device buffer sharing may require expensive data copies. Sharing the tensor requires coordinating between the respective WebNN and WebGPU timelines. Below is an example of how the user agent may coordinate this handoff:
+
+- Two fences are created:
+  1. a "start access" fence which is to be signaled by WebNN and waited on by WebGPU. A data copy may be required alongside the signaling of this fence
+  2. an "end access" fence which is to be signaled by WebGPU and waited on by WebNN. A data copy may be required alongside the signaling of this fence
+- The `GPUDevice` enqueues a command to its `GPUQueue` to wait for the "start access" fence to be signaled
```

Comment:
This line cannot be implemented on all systems, even though it would happen to work when WebML uses DML (or maybe even the NPUs on Windows not sure). On other systems, hardware / GPU drivers might not allow waiting on fences that haven't been previously signaled, so as to guarantee forward progress. This means that we cannot enqueue GPU work that waits on a fence, and only later signal it when the CPU detects that the graph execution is finished.

The other issue is that the computations here require a CPU, userspace roundtrip to signal the fence, it would be ideal if the NPU could signal a fence that immediately causes the OS to make its signaling to the GPU, but I don't know that any OS allows that (let alone that it is a portable capability that we can rely on).

The only option to implement the current semantics between a GPU and another accelerator (GPU, CPU or NPU) is to make Chromium implement the deferred submission of GPU work until WebNN is finished executing the graph. However I **strongly** recommend not doing that as a bunch of other things would need to be deferred, like whole compositor frames (so GPUCanvas can be displayed correctly) etc. That would be extremely invasive in both implementations and various HTML / canvas specs.

The only remaining possibility to implement this interface would be to make the whole `MLContext` hosted on the same GPU as the `GPUDevice`, because then everything is just GPU operations that are ordered in the same timeline. This is what would happen with a DML implementation with the difference that Windows is a system where GPU1-GPU2 fences work so DML could be on a different GPU.

Reply:
Thank you for the very important feedback!

> On other systems, hardware / GPU drivers might not allow waiting on fences that haven't been previously signaled, so as to guarantee forward progress. This means that we cannot enqueue GPU work that waits on a fence, and only later signal it when the CPU detects that the graph execution is finished.

Out of curiosity, do you know which specific platforms have this issue? Is Metal one of them?

> The only option to implement the current semantics between a GPU and another accelerator (GPU, CPU or NPU) is to make Chromium implement the deferred submission of GPU work until WebNN is finished executing the graph. However I strongly recommend not doing that as a bunch of other things would need to be deferred

I agree that would be bad. I think we have to change the semantics of the `importExternalBuffer()` (or whatever we decide to name it) method. If it was a promise-bearing async API (at least on platforms which need to wait on the CPU for the ML device to finish operations), we can wait for ML operations to complete from a background CPU thread and resolve the promise at that time.

@Kangz do you see any obvious issues with this approach? I can flesh out this idea more thoroughly tomorrow

I believe the handoff in the other direction (returning a buffer to WebNN) does not require any changes. We can delay queueing any more WebNN operations dependent (directly or indirectly) on the `MLTensor` rented to WebGPU until WebGPU releases the buffer - basically, WebNN's timeline would be dependent on WebGPU, but not the other way around. The results of a given WebNN operation are not observable until either the promise-bearing `readBuffer()` or `importExternalBuffer()` methods are called, so it shouldn't be a problem that all these WebNN operations are queued behind WebGPU work.

> The only remaining possibility to implement this interface would be to make the whole `MLContext` hosted on the same GPU as the `GPUDevice` because then everything is just GPU operations that are ordered in the same timeline

That only works on platforms which express their ML compute in terms of GPU commands, such as DirectML. See prior discussions of `MLCommandEncoder` in #255 and #303, which was abandoned.

For example, while CoreML is built on top of primitives such as [Metal Performance Shaders](https://developer.apple.com/documentation/metalperformanceshaders), it does not have a mechanism to access the underlying `MTLCommandBuffer`s which may or may not be used to execute the graph (where the model executes is a black box - see #749). There is no way to force CoreML to use a given GPU. CoreML allows you to set a [preferredMetalDevice](https://developer.apple.com/documentation/coreml/mlmodelconfiguration/preferredmetaldevice), but (1) while it seems unlikely that CoreML would choose another GPU, it’s clearly stated as a _preference_, and (2) the model may run on another [compute unit](https://developer.apple.com/documentation/coreml/mlcomputeunits) (i.e. CPU, ANE) entirely.

As far as I know, CoreML's caller must dispatched graph execution from the CPU and completion may only be awaited from CPU. @mwyrzykowski please correct me if that's not true!

Reply:
> I think we have to change the semantics of the `importExternalBuffer()` method... [to be] a promise-bearing async API

Tagging @RafaelCintron and @reillyeon since I can't take full credit for this idea :)

Reply:
> Out of curiosity, do you know which specific platforms have this issue? Is Metal one of them?

I'm not sure, I know that at least a bunch of Vulkan devices (so Linux, ChromeOS and Android) have this issue as it is only available with "timeline semaphores" that aren't universally available. I don't know if `MTLEvent` lets you wait before signaling (in your comment you mentioned CoreML can only sync with the CPU). If folks in this group are able to prove that this capability is universally available (ideally with some tests on a variety of hardware showing it) then it can rely on it.

>  If it was a promise-bearing async API (at least on platforms which need to wait on the CPU for the ML device to finish operations), we can wait for ML operations to complete from a background CPU thread and resolve the promise at that time.

A Promise would work to alleviate this issue for sure! Adding a sync API when possible is fine, but would very much decrease the portability of the API as "works on my machine" no longer means "works on everyone's machine". WebGPU has more mechanisms around optional feature to try to push developers in the portable by default direction.

> That only works on platforms which express their ML compute in terms of GPU commands, such as DirectML.

It would also work with a full implementation of WebNN using the underlying WebGPU implementation, but that obviously wouldn't use the NPU so IDK if you'd want to pursue that.

Reply:
> If folks in this group are able to prove that this capability is universally available (ideally with some tests on a variety of hardware showing it) then it can rely on it.

Other folks here are welcome to try, but now that I'm more aware of the implications I am skeptical this is a dependency we want to forever bake into the web platform... :) 

>> If it was a promise-bearing async API 

This is now officially the proposal. PTAL, thanks!

> > That only works on platforms which express their ML compute in terms of GPU commands, such as DirectML.
> 
> It would also work with a full implementation of WebNN using the underlying WebGPU implementation, but that obviously wouldn't use the NPU so IDK if you'd want to pursue that.

Yes, true. While it may be reasonable for WebNN to be implemented using WebGPU on some platforms, I think we've failed if we've backed WebNN into a corner such that _it can only be implemented on top of WebGPU (or DirectML)_

Reply:
> I'm not sure, I know that at least a bunch of Vulkan devices (so Linux, ChromeOS and Android) have this issue as it is only available with "timeline semaphores" that aren't universally available. 

Timeline semaphores are [a core feature of Vulkan 1.2](https://www.khronos.org/blog/vulkan-timeline-semaphores). Previous to that, they were available as an extension. Vulkan 1.2 is supported by all of the major IHVs so it's a question of how new your device and driver. I do not have numbers on how many devices in the wild support it.

Reply:
I chatted with @Kangz offline and will summarize our discussion here... We're concerned that supporting the sync variant of `importExternalBuffer()` on all platforms may not be feasible. That being said, on platforms which _do_ naturally support interleaving ML compute with other GPU workloads - which we assume is a desirable feature to both native apps and web apps - it would be unfortunate to not expose this capability to the web. We'd like to collect more data and implementation experience to help inform next steps.

To that end, we expect to prototype _both the sync and async variants_ of `importExternalBuffer()` in Chromium:
- the async variant as the default for all backends (starting with Windows), and
- the sync variant _only on Windows when executing WebNN graphs on the GPU_, which is the best-case scenario for this kind of interop. This will be a useful baseline for comparing performance

Reply:
Resolving for now. Please re-open if there's more to discuss